Wrap login options in accordion

emlun opened the pull request:
This is a meta-PR into:

- https://github.com/syncthing/syncthing/pull/9175

@acolomb I believe this should address your review: https://github.com/syncthing/syncthing/pull/9175#pullrequestreview-2204685097

emlun committed:
Wrap login options in accordion

emlun committed:
Move submit button and "stay logged in" checkbox next to each other

emlun committed:
Prettify margins in login panels

emlun committed:
Save most recent login option in localStorage if possible

emlun committed:
Apply primary class to login form submit button

emlun committed:
Tweak error message per review

emlun committed:
Disable WebAuthn "Stay logged in" checkbox when button is disabled

acolomb commented on gui/default/assets/css/overrides.css:
```diff
@@ -569,6 +569,14 @@ html[lang|="ko"] i {
     user-select: all;
 }
 
+#login-options .panel-body {
+  padding: 1em;
+}
+
+#login-options .stay-logged-in {
+  margin-top: 0.25em;
+}
```
Could use some comments here, this seems very specific and it's not clear why.

emlun replied:
Ok, but I don't know what would be useful to put in such a comment. It just looks better to me this way? (and I didn't find any suitable pre-baked spacing classes in [Bootstrap CSS](https://getbootstrap.com/docs/3.3/css/))

acolomb replied:
The comment should state where this override applies.  I have no idea what it does, as the CSS classifiers don't speak for themselves and it would be better not having to search the HTML code to see where this applies.

emlun replied:
Ok. This made me look more into why the padding style is needed, and I discovered [this global padding decrease class](https://github.com/syncthing/syncthing/blob/13d545c9f2e54ac32e35047a475a3f4bbd616e70/gui/default/assets/css/overrides.css#L264-L266). Since that style is no longer appropriate for all collapsible panels, I renamed it to `.panel-body.less-padding` and added the `less-padding` class everywhere it currently applies. Hopefully that name should be self-explanatory enough that its comment is no longer needed.

Also added a comment to the remaining `#login-options .stay-logged-in` style in 709cbc121bae3c0015b134e081d01d2e1291e516.

acolomb commented on gui/default/index.html:
```diff
@@ -344,118 +344,190 @@ <h3 class="panel-title">
+              data-parent="#login-options"
+              data-target="#login-option-password"
+              aria-expanded="{{isDefaultLoginOption('password') ? 'true' : 'false'}}"
+            >
```
This looks very different from the HTML code style used throughout the rest of the file.  I'd err on the side of consistency, although for sure the sometimes very long lines are not very readable.  And it does add quite a bit of spaces, needlessly inflating the file size.  Of course it's not much, but avoidable.  And it makes the diffs harder to parse.

emlun replied:
I emphatically disagree with that order of priorities, but ok, done in c393ab3ff131f3d64511cf8f5d4c9139548ab9b4. And to be fair I noticed that I did also line-break a couple of tags that didn't otherwise change (apart from indentation) - my bad, that didn't belong in this diff either way (or at the very least, it should have been its own commit).

> And it does add quite a bit of spaces, needlessly inflating the file size.

May I ask the reason for this concern? Because if you're worried about the size of the Git repo, I don't think you should be - Git by default compresses content in packfiles using zlib, and indentation compresses _very_ well (default `zlib-flate` settings compress this file from 80,513B to 11,316B). Likewise for keeping the file small over the network, I think one should make sure to use HTTP compression (which would make micro-optimizations like this wholly insignificant) before worrying about source file size.

Anyway, done in c393ab3ff131f3d64511cf8f5d4c9139548ab9b4.

acolomb replied:
Sure, these are opinionated and subjective details.  I just want the changes to cause the minimum amount of friction, which is harder if unrelated style changes are part of them.

I don't really care about the repo size, no.  But the GUI is served from the Syncthing backend to a browser, and that connection might be slow, so we should try not to be wasteful without a very good reason.  I'm not sure whether compressed encoding is supported there.  I've seen some slow and flaky network connections, and it really helps to keep the files smaller there.  But that is actually the lowest priority for me.

1. Efficiency (objective).  At such a level of indentation, each line adds a number of indent spaces.  If the text after that is not significantly longer than the spaces, we may be wasting more storage for garbage than actual content.  That's when I start to reconsider whether it's really necessary.

2. Diff-friendly.  Breaking lines often doesn't help in that regard.  Yes, HTML is already a terrible language to read and diff, even when ignoring indentation-only changes.  So let's try not to make it harder.

3. Translations.  As stated above, `(space)Text(space)` behaves differently within an HTML element than without the spaces.  For automated translations, I think the strings are trimmed, which means that the translated strings might wrap differently from the original source.  That's a can of worms we don't need.

4. Most important, blend in.  The surrounding code follows a certain style pattern and it helps reading and finding stuff if that is used consistently.  It might not match everyone's personal preferences, but when many people touch the code, the end result should still look like it was written as a whole.  For now, that's pretty long lines where many attributes are needed.

Thanks for accepting these nit-picks.  Again, in the end I'm just trying to reduce friction by pointing out these patterns so the remaining review can concentrate on the functionality.

emlun replied:
1. Again, I think HTTP compression would be the proper solution for that; [MDN claims](https://developer.mozilla.org/en-US/docs/Web/HTTP/Compression) that it's ubiquitously supported in servers and browsers.
2. I agree with the goal, but I'm of the opposite opinion that breaking long lines _does_ help diff-friendliness, and _a lot_ at that. But that is of course assuming that the line breaks were already there, before changing the substance, and as I said I admit and agree that I failed on that part here (and even worse as I introduced some line breaks with no change in substance).
3. Ah, I see. I hadn't considered that.
4. Yeah, on some level I agree, but I would still advocate for improving code style incrementally as and where changes are made, rather than condemning the entire file to forever remain in a style that harms readability and diff-friendliness. If the choice is between (1) never change style, (2) change all the style at once, and (3) change style incrementally with new changes, I think (3) is by far the best. (1) makes the code artificially hard to work with; (2) tends to be postponed forever, and if/when you do make the leap, it becomes a "history barrier" that makes change history very hard to track beyond that point (and also affects merges and rebases), in a way that (3) does not.

Anyway, sorry, I'm not really trying to convince you to change your mind, just wanted to share a different perspective. I do appreciate these comments and consider these "nitpicks" no less worthy than other comments.

acolomb commented on gui/default/index.html:
```diff
@@ -344,118 +344,190 @@ <h3 class="panel-title">
+              <h4 class="panel-title">
+                <div class="panel-title-text" translate>
+                  Password
+                </div>
```
Again, please try to shorten and adhere to the surrounding style.  And do note that there is a difference between having whitespace inside an element or not.  Which sometimes gives very strange behavior w.r.t. wrapping.

emlun replied:
Done in dbe192ec8e40baaaba523cdb7c2d92885b8674c4 (as far as I can tell).

acolomb reviewed (commented):
Yeah, I like this a lot better.  Could use some more spacing between the two options.  Maybe even put a paragraph with a big fat `--- or ---` in between.

Comments are mostly about the HTML coding style.  Please try to keep the diffs small and not needlessly reformat / wrap HTML code.  I know it's hard to read as it is, but the diff with your wrapping changes just makes it much harder to see what's changing.

acolomb commented:
Note that I haven't actually tested any WebAuthn logins yet.  But code-wise it looks like a nice a clean solution with the local storage approach.

emlun committed:
Line-join tags with short content

emlun committed:
Line-join HTML tags

emlun commented:
> Maybe even put a paragraph with a big fat `--- or ---` in between.

I tried this:

![image](https://github.com/user-attachments/assets/85c68fcb-a596-4500-8986-076969ab34c6)

Personally I don't like it, I much prefer the look with the plain panels:

![image](https://github.com/user-attachments/assets/b1755e8a-e665-4cb5-bd30-84fd957c47c2)

I also worry that the "OR" might get a bit janky when translated. For example in English, "or" and "Or" look visually unbalanced, so it should be "OR", so I put `text-transform: uppercase` for that sake. But in Swedish I would write "--- Eller ---" much rather than "--- eller ---" or "--- ELLER ---". So this would need a bespoke context-aware translation key just for this occurrence to account for typographic quirks like that.

Also, this particular HTML solution depends on flexbox, which I'm not sure we assume is supported by every user's browser? This would be very finnicky to make without flexbox, so I'd much prefer to just scrap it.


> Could use some more spacing between the two options.

I've reused the same panel component as in the main GUI - should we really introduce this spacing difference between those uses?

acolomb commented:
Yes, I think a difference in spacing is justified.  In the main GUI, it's a list (repetition) of the same type of thing.  Here, it's a contrast between two alternatives.

The horizontal line looks fine to me with its spacing.  We could just leave out the "OR" text and only have the line.  I agree that it's a headache in translations.

Sorry, I don't know what flexbox means and what the exact usage is.  But adding some space via a CSS override should be easy and portable.  Just as adding a `<hr>` in between.

acolomb commented on gui/default/index.html:
```diff
@@ -344,118 +344,142 @@ <h3 class="panel-title">
+                      </label>
+                    </div>
+                    <div class="col-md-3 text-right">
+                      <button type="submit" id="submit" class="btn btn-primary" ng-disabled="login.inProgress" translate>
+                        Log In
+                      </button>
```
No wrap, as it was on main.

emlun replied:
Thanks, done in 00ba31eb965e286e0f1c468cd8c04d83cbb8f3d9.

acolomb commented on gui/default/index.html:
```diff
@@ -344,118 +344,142 @@ <h3 class="panel-title">
+                  <div class="col-md-6">
+                    <label class="stay-logged-in">
+                      <input type="checkbox" ng-model="login.stayLoggedIn" ng-disabled="login.inProgress || isWebauthnLoginDisabled()">
+                      &nbsp;
```
An `&nbsp;` on a line of its own is useless.  Because the line break and indentation spaces do count as a single regular space, so wrapping before or after your non-breaking space is still done.  See the `<!-- Avoid stray space...` comments in this file for an example how to avoid that, while still wrapping lines for readability.

emlun replied:
Good catch, thanks. Done in 00ba31eb965e286e0f1c468cd8c04d83cbb8f3d9.

acolomb reviewed (commented):
Great job overall, I think with the additional spacing this will look awesome.  Just two more small comments regarding HTML whitespace.

emlun committed:
Line-join more HTML tags

emlun committed:
Add divider line between login option panels

emlun committed:
Replace global "Panel padding decrease" style with targeted class

This padding decrease currently applies to _all_ collapsible panels, but this
padding decrease is not appropriate for all collapsible panels.

emlun committed:
Replace global "Panel padding decrease" style with targeted class

This padding decrease currently applies to _all_ collapsible panels, but this
padding decrease is not appropriate for all collapsible panels.

emlun committed:
Remove .less-padding from login form panels

emlun committed:
Explain #login-options .stay-logged-in style

acolomb commented on gui/default/assets/css/overrides.css:
```diff
@@ -569,10 +565,7 @@ html[lang|="ko"] i {
-  padding: 1em;
-}
-
+/* Align checkbox with button in login forms */
```
Ah, I see.  Maybe there's a better way to align them by centering vertically in some common container?  The Sharing tab when editing a folder / device wraps a checkbox and a text entry into a common `<div class="form-group">`, maybe that would be an easier solution here as well?

emlun replied:
Nope, these elements are already wrapped with `.form-group`, and the only thing that class does is `margin-bottom: 15px;`.

The correct solution to this would be to use a [flexbox](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_flexible_box_layout/Basic_concepts_of_flexbox) with `align-items: baseline` to align the text baselines of the checkbox label and the button. But flexbox is somewhat new, and I've seen some concerns about compatibility with older browsers, so I didn't reach for that solution for now. With your blessing I would be happy to do that, though.

acolomb replied:
Shouldn't this be really simple? https://stackoverflow.com/a/7311753

emlun replied:
Unfortunately no...

`display: inline-block` and `vertical-align: middle` works for getting them aligned vertically, but breaks the horizontal right-alignment of the button. The right-alignment is currently achieved using the `col-md-9` and `col-md-3` classes, which set the `float` property which forces `display:` to `block`. I tried a couple other things but didn't get both vertical alignment and button-right-alignment to work simultaneously.

Actually, flexbox has been [widely supported since 2015](https://caniuse.com/flexbox), so maybe we should just use that after all. In browsers that support it, it'll look like this:

![image](https://github.com/user-attachments/assets/671e416c-c71b-4be0-834a-0d15d59e96cc)

And in browsers that don't support it, it'll be ugly but still functional:

![image](https://github.com/user-attachments/assets/bc16d923-24b5-4309-ad0a-bab341090799)

Does that seem acceptable?

acolomb replied:
Actually I think having the button on the right side is more important than the vertical alignment.

Taking a step back, why don't we just put the button on a row of its own, as in all other dialogs.  It's just the utter desire of utilizing the empty space left of the button for the checkbox that gets us into this trouble.  Everywhere else, the action buttons are alone on their row and we never have these problems.  Let's try that and see what ideas others come up with during review of the whole PR.  Acceptable?

As in: The flexbox thing would be usable and maybe beneficial in a lot of places.  Let's not make the decision for this panel alone, but if we decide to adopt the new technology, then for the whole GUI layout preferably.  Otherwise that might hold up the WebAuthn PR further with such side-track discussions.

emlun replied:
Ok, that sounds fair.

acolomb commented on gui/default/assets/css/overrides.css:
```diff
@@ -312,11 +312,7 @@ a.toggler:hover {
- */
-
-.panel-collapse .panel-body {
+.panel-body.less-padding {
```
This change should probably be separated out into a new PR on the main repo.  It touches many files and makes the PR needlessly bigger.

emlun replied:
Ok, see https://github.com/syncthing/syncthing/pull/9659

acolomb commented on gui/default/index.html:
```diff
@@ -344,118 +344,140 @@ <h3 class="panel-title">
+                <div class="form-group">
+                  <div class="col-md-6">
+                    <label class="stay-logged-in">
+                      <input type="checkbox" ng-model="login.stayLoggedIn" ng-disabled="login.inProgress || isWebauthnLoginDisabled()">&nbsp;<span translate>Stay logged in</span>
```
Actually, I find it surprising that this checkbox is locked.  Since it's synchronized with the other one, I see no reason to block its usage just because the login method is unavailable.

emlun replied:
Ok, fixed in abab9e91cd963a49cdc9916b5ca93d41c9d9b89d and 292a9edced589b5d1e01049646515fcc623a450e.

acolomb commented on gui/default/index.html:
```diff
@@ -344,118 +344,140 @@ <h3 class="panel-title">
+                  <div class="form-group">
+                    <div class="col-md-9">
+                      <label class="stay-logged-in">
+                        <input type="checkbox" ng-model="login.stayLoggedIn" >&nbsp;<span translate>Stay logged in</span>
```
Shouldn't this be disabled while the login is in progress, like the other one?

emlun replied:
Sure, fixed in b289e5279e1b0d9a3bf2ce1a689db1e14fcfedba.

emlun committed:
Merge branch 'webauthn' into webauthn-login-page

emlun committed:
Merge branch 'less-padding-class' into webauthn-login-page

emlun committed:
Revert "Disable WebAuthn "Stay logged in" checkbox when button is disabled"

This reverts commit 0a2e6bcf56a74a28f466a47642202f4cc3799044.

emlun committed:
Disable "stay logged in" checkboxes while login is in progress

emlun committed:
Set login.inProgress state during WebAuthn login

emlun committed:
Draw "stay logged in" checkbox and "log in" button on separate rows

emlun committed:
Remove empty else statement

acolomb commented:
Actually, one more try at the alignment stuff... This is what it would look like if we add a panel-footer section and move the button and the checkbox into that:

![image](https://github.com/user-attachments/assets/6deb79a7-0b4f-42e6-bd22-45736c35e9f9)

Feel free to ignore or holler if you want me to propose the code changes if you like the look.

emlun commented:
The text is still not quite aligned correctly:

![valign](https://github.com/user-attachments/assets/27827350-4c96-4b72-b43e-f9f17cfad31a)

But sure, you're very welcome to submit that as a meta-PR into https://github.com/syncthing/syncthing/pull/9175 if you prefer the look.
